src/Microsoft.OData.Client/ALinq/DataServiceQueryProvider.cs

```diff
@@ -109,11 +113,94 @@ internal TElement ReturnSingleton<TElement>(Expression expression)
+                    case SequenceMethod.AverageSingleSelector:
+                    case SequenceMethod.AverageNullableIntSelector:
+                    case SequenceMethod.AverageNullableDoubleSelector:
+                    case SequenceMethod.AverageNullableDecimalSelector:
+                    case SequenceMethod.AverageNullableLongSelector:
+                    case SequenceMethod.AverageNullableSingleSelector:
+                    case SequenceMethod.MinSelector: // Mapped to a generic expression - Min(IQueryable`1<T0>, Expression`1<Func`2<T0, T1>>)->T1
+                    case SequenceMethod.MaxSelector: // Mapped to a generic expression - Max(IQueryable`1<T0>, Expression`1<Func`2<T0, T1>>)->T1
+                        {
+                            Func<QueryResult, TElement> parseQueryResultFunc = (queryResult) =>
+                            {
+                                IDictionary<string, string> responseHeaders = new Dictionary<string, string>();
+                                responseHeaders.Add(ODataConstants.ContentTypeHeader, "application/json");
+                                HttpWebResponseMessage httpWebResponseMessage = new HttpWebResponseMessage(responseHeaders, (int)queryResult.StatusCode, queryResult.GetResponseStream);
+
+                                ODataMessageReaderSettings messageReaderSettings = new ODataMessageReaderSettings
+                                {
+                                    Validations = ~ValidationKinds.ThrowOnUndeclaredPropertyForNonOpenType
+                                };
+
+                                ODataResource entry = default(ODataResource);
+                                using (var messageReader = new ODataMessageReader(httpWebResponseMessage, messageReaderSettings, this.Context.Format.ServiceModel))
+                                {
+                                    var reader = messageReader.CreateODataResourceSetReader();
+                                    while (reader.Read())
+                                    {
+                                        switch (reader.State)
+                                        {
+                                            case ODataReaderState.ResourceEnd:
+                                                entry = reader.Item as ODataResource;
+                                                if (entry != null && entry.Properties.Any())
+                                                {
+                                                    ODataProperty aggregationProperty = entry.Properties.First();
+                                                    ODataUntypedValue untypedValue = aggregationProperty.Value as ODataUntypedValue;
+
+                                                    Type underlyingType = Nullable.GetUnderlyingType(typeof(TElement));
+                                                    if (underlyingType == null) // Not a nullable type
+                                                    {

```

Comment:
could this be a stand-alone function maybe?

Reply:
@wachugamaina Fixed